Bad codegen when casting procs

```nim
proc foo() {.importc.}
(cast[proc(a: int) {.cdecl.}](foo))(5)
```
```
Error: execution of an external compiler program 'clang -c  -w  -I/Users/yglukhov/Projects/nim/lib -o /Users/yglukhov/nimcache/test2.o /Users/yglukhov/nimcache/test2.c' failed with exit code: 1

/Users/yglukhov/nimcache/test2.c:94:6: error: too many arguments to function call, expected 0, have 1
        foo(((NI) 5));
        ~~~ ^~~~~~~~
/Users/yglukhov/nimcache/test2.c:13:11: note: 'foo' declared here
N_NIMCALL(void, foo)(void);
          ^
1 error generated.
```

--- a/compiler/semfold.nim
+++ b/compiler/semfold.nim
@@ -771,7 +771,7 @@ proc getConstExpr(m: PSym, n: PNode; idgen: IdGenerator; g: ModuleGraph): PNode 
   of nkCast:
     var a = getConstExpr(m, n[1], idgen, g)
     if a == nil: return
-    if n.typ != nil and n.typ.kind in NilableTypes:
+    if n.typ != nil and n.typ.kind in NilableTypes-{tyProc}:
       # we allow compile-time 'cast' for pointer types:
       result = a
       result.typ = n.typ

--- a/compiler/semfold.nim
+++ b/compiler/semfold.nim
@@ -771,7 +771,8 @@ proc getConstExpr(m: PSym, n: PNode; idgen: IdGenerator; g: ModuleGraph): PNode 
   of nkCast:
     var a = getConstExpr(m, n[1], idgen, g)
     if a == nil: return
-    if n.typ != nil and n.typ.kind in NilableTypes-{tyProc}:
+    if n.typ != nil and n.typ.kind in NilableTypes and
+        not (n.typ.kind == tyProc and a.typ.kind == tyProc):
       # we allow compile-time 'cast' for pointer types:
       result = a
       result.typ = n.typ

--- a/tests/misc/tcast.nim
+++ b/tests/misc/tcast.nim
@@ -1,6 +1,7 @@
 discard """
   output: '''
 Hello World
+Hello World
 Hello World'''
   joinable: false
 """
@@ -8,7 +9,7 @@ type MyProc = proc() {.cdecl.}
 type MyProc2 = proc() {.nimcall.}
 type MyProc3 = proc() #{.closure.} is implicit
 
-proc testProc()  = echo "Hello World"
+proc testProc() {.exportc:"foo".} = echo "Hello World"
 
 template reject(x) = doAssert(not compiles(x))
 
@@ -22,6 +23,10 @@ proc callPointer(p: pointer) =
 
   ffunc0()
   ffunc1()
+
+    # bug #5901
+  proc foo() {.importc.}
+  (cast[proc(a: int) {.cdecl.}](foo))(5)
 
 callPointer(cast[pointer](testProc))
 

